lib/improver/nowcasting/optical_flow.py

```diff
@@ -0,0 +1,830 @@
+                                             ygrid, xpt, ypt, xwt, ywt)
+
+        return adv_field
+
+    def process(self, cube, timestep, fill_value=0.0):
+        """
+        Extrapolates input cube data and updates validity time.  The input
+        cube should have precisely two non-scalar dimension coordinates
+        (spatial x/y), and is expected to be in a projection such that grid
+        spacing is the same (or very close) at all points within the spatial
+        domain.  The input cube should also have a "time" coordinate.
+
+        Args:
+            cube (iris.cube.Cube):
+                The 2D cube containing data to be advected
+            timestep (datetime.timedelta):
+                Advection time step
+            fill_value (float):
+                Default output value for spatial points where data cannot be
+                extrapolated (source is out of bounds)
+
+        Returns:
+            advected_cube (iris.cube.Cube):
+                New cube with updated time and extrapolated data
+        """
+        # check that the input cube has precisely two non-scalar dimension
+        # coordinates (spatial x/y) and a scalar time coordinate
+        check_input_coords(cube, require_time=True)
+
+        # check spatial coordinates match those of plugin velocities
+        if (cube.coord(axis="x") != self.x_coord or
+                cube.coord(axis="y") != self.y_coord):
+            raise InvalidCubeError("Input data grid does not match advection "
+                                   "velocities")
+
+        # derive velocities in "grid squares per second"
+        def grid_spacing(coord):
+            """Calculate grid spacing along a given spatial axis"""
+            new_coord = coord.copy()
+            new_coord.convert_units('m')
+            return float(np.diff((new_coord).points)[0])
+
+        grid_vel_x = self.vel_x.data / grid_spacing(cube.coord(axis="x"))
+        grid_vel_y = self.vel_y.data / grid_spacing(cube.coord(axis="y"))
+
+        # perform advection and create output cube
+        advected_data = self._advect_field(cube.data, grid_vel_x, grid_vel_y,
+                                           timestep.total_seconds(),
+                                           fill_value)
+        advected_cube = cube.copy(data=advected_data)
+
+        # increment output cube time
+        original_datetime, = \
+            (cube.coord("time").units).num2date(cube.coord("time").points)
+        new_datetime = original_datetime + timestep
+        new_time = (cube.coord("time").units).date2num(new_datetime)
+        advected_cube.coord("time").points = new_time
+
+        return advected_cube
+
+
+class OpticalFlow(object):
+    """
+    Class to calculate advection velocities along two orthogonal spatial axes
+    from time-separated fields using an optical flow algorithm
+    """
+
+    def __init__(self, kernel=7, smethod='box', boxsize=30, point_weight=0.1,
```

Comment:
kernel_size and smoothing_method? Makes it clearer as these will be exposed, sometimes out of context.